### PR TITLE
Do not omit the else part when an unswitched loop contains barriers.

### LINKED ISSUE #862: Wrong codegen results lead to silent out of sync blockBroadcast that produces wrong result
## 🐛 Bug

Have discussed this offline with @naoyam 

I'm observing a generated code like the one below, where we have a broadcast::blockBroadcast inside a predicate that is missing the else block.
I believe this leads to an out of sync broadcast down the line that is giving me wrong results.

```
119   #pragma unroll 1                                                                                                      
120   for(size_t ki342 = 0; ki342 < (ceilDiv((ceilDiv(16, 1)), 1)); ++ki342) {                                              
121     if ((((((((((ki342 * 1) + (1 - 1)) * 1) + (1 - 1)) * (ceilDiv((T0.size[0] * (T0.size[2] * T0.size[3])), 16))) + threadIdx.x) / (T0.size[2] * T0.size[3])) < T0.size[0]) && ((((((((ki34    2 * 1) + (1 - 1)) * 1) + (1 - 1)) * (ceilDiv((T0.size[0] * (T0.size[2] * T0.size[3])), 16))) + threadIdx.x) % (T0.size[2] * T0.size[3])) < (T0.size[2] * T0.size[3])))) {
122       constexpr int64_t ki344 = 0;                                                                                      
123       constexpr int64_t ki346 = 0;                                                                                      
124       float T4[1];                                                                                                      
125       broadcast::blockBroadcast<true, false, false>(                                                                    
126         T4[0],                                                                                                          
127         T3[0],                                                                                                          
128         static_cast<float*>(shared_mem),                                                                                
129         true);                                                                                                          
130       T5[((((ki342 * 1) + ki344) * 1) + ki346)]                                                                         
131         = T1[((((ki342 * 1) + ki344) * 1) + ki346)]                                                                     
132         - T4[0];                                                                                                                                                                           
133     }                                                                                                                   
134   }     
```

Current WAR (thanks to Naoya for pointing out) is to disable the `canOmitElseClause` at : https://github.com/csarofeen/pytorch/blob/93f25141c3d8d02d8e96012b089c6ead4d25fc7e/torch/csrc/jit/codegen/cuda/lower_unroll.cpp#L206

## To Reproduce

I have pushed the code to: https://github.com/csarofeen/pytorch/tree/sync_bug_else_block
Build and run with
`./build/bin/test_jit --gtest_filter="BNRepro2"`

### PR DESCRIPTION
**Pull request by naoyam**
Fixes #862 

Reproducer, `FusionBNRepro2`,  by @jjsjann123

### CODE REVIEW

File: test/cpp/jit/test_gpu.cpp [hunk 1]
```diff
@@ -14957,6 +14957,91 @@ TEST(NVFuserTest, FusionBNRepro_CUDA) {
...
+  // This works, so we can use it as a reference
+  int h = 8;
+  int w = 8;
+#endif
+  int numDims = 4;
+
+  auto input = makeSymbolicTensor(numDims);
+  fusion.addInput(input);
+
+  // TODO: error set 1, runtime momentum;
```
Comment (rdspring1) [inline test/cpp/jit/test_gpu.cpp]: What does this mean?

File: test/cpp/jit/test_gpu.cpp [hunk 2]
```diff
@@ -14957,6 +14957,91 @@ TEST(NVFuserTest, FusionBNRepro_CUDA) {
...
 }
 
+TEST(NVFuserTest, FusionBNRepro2_CUDA) {
+  std::unique_ptr<Fusion> fusion_ptr = std::make_unique<Fusion>();
+  Fusion& fusion = *fusion_ptr.get();
+  FusionGuard fg(&fusion);
+
+  int batch = 2;
+  int c = 4;
+#if true
```
Comment (rdspring1) [inline test/cpp/jit/test_gpu.cpp]: Do we still need both cases?

File: torch/csrc/jit/codegen/cuda/lower_unroll.cpp [hunk 3]
```diff
@@ -206,22 +206,64 @@ void UnrollPass::handle(kir::ForLoop* fl) {
...
+        }
+      }
+    }
+    // If the number of visits of the loop body per thread is one, the
+    // unswitch predicate is sufficient.
+    // When the loop stop is the same as the extent of its IterDomain,
+    // the per-thread visit count is guaranteed to be one at most (see
+    // CudaKernelGenerator::visit(kir::ForLoop*) as well. Also, when a
+    // loop is vectorized (not misaligned), the count must be one at
+    // omst. Even if not parallelized nor vectoirzed, it is also
```
Comment (rdspring1) [inline torch/csrc/jit/codegen/cuda/lower_unroll.cpp]: 'omst' => 'most'
'vectoirzed' => 'vectorized'

Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/lower_unroll.cpp]: typo `omst` -> `most`

File: torch/csrc/jit/codegen/cuda/lower_unroll.cpp [hunk 4]
```diff
@@ -206,22 +206,64 @@ void UnrollPass::handle(kir::ForLoop* fl) {
...
   std::vector<kir::ForLoop*> loops({fl});
+
   while (loops.size() > 0) {
     auto loop = loops.back();
     loops.pop_back();
+
+    // If there's any expression that requires barrier
+    // synchronization, the else part can't be omitted
+    for (auto expr : loop->body().exprs()) {
+      if (auto broadcast = dynamic_cast<kir::BroadcastOp*>(expr)) {
```
Comment (rdspring1) [inline torch/csrc/jit/codegen/cuda/lower_unroll.cpp]: `broadcast` isn't used in the If-block, so would `expr->isA<kir::BroadcastOp*>()` be more appropriate?

**Review by jjsjann123 — APPROVED** [review]
LGTM

File: torch/csrc/jit/codegen/cuda/lower_unroll.cpp (hunk 3 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_unroll.cpp]: Thanks. Fixed.

File: torch/csrc/jit/codegen/cuda/lower_unroll.cpp [hunk 5]
```diff
@@ -206,22 +206,64 @@ void UnrollPass::handle(kir::ForLoop* fl) {
...
+        const ParallelTypeBitmap domains =
+            ir_utils::getParallelBroadcastDomains(
+                expr->outputs()[0]->as<kir::TensorView>()->fuserTv(),
```
Comment (rdspring1) [inline torch/csrc/jit/codegen/cuda/lower_unroll.cpp:220]: Should we use `firstTvOutput` in predicate_compute.cpp here to replace `expr->outputs()[0]->as<kir::TensorView>()?` It is the kir version of `getTVOutput` in lower_utils.cpp.

File: torch/csrc/jit/codegen/cuda/lower_unroll.cpp [hunk 6]
```diff
@@ -206,22 +206,64 @@ void UnrollPass::handle(kir::ForLoop* fl) {
...
+                thread_predicates_);
+        if (domains.any()) {
+          return false;
+        }
+      } else if (expr->isA<kir::ReductionOp>() || expr->isA<kir::WelfordOp>()) {
+        auto td = expr->outputs()[0]->as<kir::TensorView>()->domain();
```
Comment (rdspring1) [inline torch/csrc/jit/codegen/cuda/lower_unroll.cpp]: Maybe replace `expr->outputs()[0]->as<kir::TensorView>()` with `firstTvOutput`.

**Review by rdspring1 — APPROVED** [review]
LGTM. I left minor comments.

File: torch/csrc/jit/codegen/cuda/lower_unroll.cpp (hunk 4 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_unroll.cpp]: Agree. Updated.

File: test/cpp/jit/test_gpu.cpp (hunk 1 above)
Comment (jjsjann123) [reply test/cpp/jit/test_gpu.cpp]: Sorry, left over code from the copy source.

**Pull request merged**